File: java/tea-framework/src/main/java/com/tea/framework/parser/SourceScanner.java
package com.tea.framework.parser;

import java.util.Arrays;
import java.util.Stack;

public class SourceScanner {

    private String buffer;
    private String[] bufferArray;
    private int position;
    private Stack<Integer> positionStack = new Stack<>();

    public int getPosition() {
        return position;
    }

    public SourceScanner(String source) {
        buffer = source;
        bufferArray = Arrays.stream(buffer.split(" "))
                .filter(b -> !b.isEmpty())
                .toArray(String[]::new);
    }

    public boolean atEndOfSource() {
        return position >= bufferArray.length;
    }

    public String read() {
        return (atEndOfSource()) ? null : bufferArray[position++];
    }

    /*
     *peeks at the next character in the buffer without incrementing the position.
     */
    public String peek() {
        return position >= bufferArray.length ? null : bufferArray[getPosition()];
    }

    public void push() {
        positionStack.push(getPosition());
    }

    public void pop() {
        position = positionStack.pop();
    }
}

File: dotnet-core/Tea/Parsing/TeaParser/SourceScanner.cs
using System;
using System.Collections.Generic;
namespace TeaParser
{
    public class SourceScanner
    {
        // the expression to parse
        readonly string? _buffer;
        readonly string[] _bufferArray;
        //
        readonly Stack<int> _positionStack = new Stack<int>();
        // the current position within the buffer.
        public int Position {get; set;}
        // are we at the end of the source?
        public bool AtEndOfSource => Position >= _bufferArray.Length;
        public SourceScanner(string source){
            _buffer = source;
            string[] splitOn = new String[] {" "};
            _bufferArray = _buffer.Split(splitOn, StringSplitOptions.RemoveEmptyEntries);
        }
        // Reads the next character in the buffer.
        public string Read()
        {
            if (AtEndOfSource)
            {
                return null;

            } else
            {
                return _bufferArray[Position++];
            }
            
        }
        ///<summary>
        /// peeks at the next character in the buffer without incrementing the position.
        ///</summary>
        public string Peek() => Position >= _bufferArray.Length ? null:  _bufferArray[Position];

        public void Push() => _positionStack.Push(Position);

        public void Pop() => Position = _positionStack.Pop();
    }
}